### PR TITLE
Call GC malloc atomic to allocate memory for primitive arrays

### PR DESCRIPTION
**Pull request by fbertra**
GC Boehm is a conservative garbage collector.  For the case where the caller knows that he won't store pointer in the allocated memory, GC Boehm provides the GC_malloc_atomic() function.  Scala primitive arrays don't store pointers, so they are candidate to call GC_malloc_atomic.

Note 1:

This PR don't cover other Scala-Native classes potentially compatible with GC_malloc_atomic allocation, like:

`
@struct class Vec (val x : Double, val y : Double, val z : Double)
`

Note 2: 

Array[Ptr[_]] are implemented as ObjectArray.

Note 3:

This app test that Boehm GC works as expected with integer arrays https://github.com/fbertra/scala-native-test-gc_malloc_atomic_primitive_arrays

This app test that Boehm GC works as expected with Array[Ptr[_]]: https://github.com/fbertra/sn-test-gc_malloc_ptr_arrays

Note 4:

Both test apps share the same problem: Boehm GC is highly random between program executions.  It is necessary to repeat execution until success or failure is detected.  This third app shows the randomness problem is independent of this PR (don't use of Array at all): https://github.com/fbertra/sn-test-boehm

### CODE REVIEW

File: nativelib/src/main/scala/scala/scalanative/runtime/GC.scala [hunk 1]
```diff
@@ -3,10 +3,17 @@ package runtime
...
+ * The Boehm GC conservative garbage collector
+ *
+ * @see [[http://hboehm.info/gc/gcinterface.html C Interface]]
+ */
 @link("gc")
 @extern object GC {
   @name("GC_malloc")
   def malloc(size: CSize): Ptr[_] = extern
+  @name("GC_malloc_atomic")
+  def malloc_pointer_free(size: CSize): Ptr[_] = extern
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/GC.scala]: I think that we want to preserve original name without `GC_` prefix here.

File: nativelib/src/main/scala/scala/scalanative/runtime/package.scala [hunk 2]
```diff
@@ -14,6 +14,19 @@ package object runtime {
     ptr
   }
 
+  /** 
+   * Allocate memory in gc heap using given info pointer.
+   *
+   * The allocated memory cannot be used to store pointers.
+   */
+  def allocPointerFree(info: Ptr[_], size: CSize): Ptr[_] = {
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/package.scala]: Best to rename to `allocAtomic` for consistency with Boehm GC.

File: nativelib/src/main/scala/scala/scalanative/runtime/ClibFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import native._
+
+/**
+ * C stdlib functions needed by nativelib
+ *
+ * Note: clib depends on nativelib.
+ */
+
+@extern
+private [scalanative] object clib_string {
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/ClibFunctions.scala]: Just `clib`.

File: nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb [hunk 4]
```diff
@@ -2,12 +2,12 @@ package scala.scalanative
...
 // Note:
-// Arrays.scala is currently implemented textual templating that is expanded through project/gyb.py script. 
-// Update the Arrays.scala.gyb and re-generate the source
+// Arrays.scala is currently implemented as textual templating that is expanded through project/gyb.py script. 
+// Update Arrays.scala.gyb and re-generate the source
 
 %{
    types = ['Boolean', 'Char', 'Byte', 'Short',
-            'Int', 'Long', 'Float', 'Double', 'Object']
+            'Int', 'Long', 'Float', 'Double']
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb]: I'd call those `primitiveTypes` now, given that `Object` is out.

File: nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb [hunk 5]
```diff
@@ -66,7 +111,8 @@ object ${T}Array {
   def alloc(length: Int): ${T}Array = {
     val arrinfo = infoof[${T}Array]
     val arrsize = sizeof[ArrayHeader] + sizeof[${T}] * length
-    val arr = runtime.alloc(arrinfo, arrsize)    
+    // pointer free -> use runtime.allocPointerFree
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb]: This comment is redundant.

File: nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb [hunk 6]
```diff
@@ -31,6 +31,51 @@ sealed abstract class Array[T]
   protected override def clone(): Array[T] = undefined
 }
 
+// note: Array[Ptr[_]] is implemented as a ObjectArray
+
+final class ObjectArray private () extends Array[Object] {
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb]: The duplication of code is unfortunate, but there is no easy way to fix it at the moment.

**Comment by densh** [comment]
Address suggested changes, reformat, squash and it's good to go!

File: nativelib/src/main/scala/scala/scalanative/runtime/ClibFunctions.scala (hunk 3 above)
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/ClibFunctions.scala]: In fact there is a better solution: we should use [LLVM's memset](http://llvm.org/docs/LangRef.html#llvm-memset-intrinsics) here instead.

Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/runtime/ClibFunctions.scala]: You can put the binding into https://github.com/scala-native/scala-native/blob/master/nativelib/src/main/scala/scala/scalanative/runtime/Math.scala and rename the file & object to `Intrinsics` instead.

**Comment by densh** [comment]
I've opened the issue for Note 1: https://github.com/scala-native/scala-native/issues/141

**Comment by densh** [comment]
Please rebase on top of latest master due to travis configuration changes.

**Comment by fbertra** [comment]
All improvements included, and more...I hope

Some notes on the new version

Note 1:

Let me argument on using allocPointerFree instead of allocAtomic.

- the long term strategy for GC is to migrate to another solution.  So, I tried to avoid using name specific to Boehm GC. 
- In JVM land, the term atomic is used in the context of low level multi-threading. It might be confusing for Scala-JVM developer (it was for me)
- I prefer allocPointerFree because it reflects better the intent.

Note 2:

The new method allocAtomicAndInit() in scalanative/runtime/package.scala is a workaround for an issue I will report soon.

Note 3:

If you like the format of the comments inside Instrinsics.scala.  I can do the same for the other instrinsics functions.

Every comment include:

- The first phrase from the overview section of the LLVM documentation page,
- a link to more information inside LLVM page.

I'm not a lawyer, but I'm pretty sure this is at least fair use!

**Comment by fbertra** [comment]
Note 3++

scalafmt didn't like my comments in Intrinsics.scala so I remove then all.

You can view how I propose to comment this file at https://github.com/fbertra/scala-native/blob/ad92ef3d2a4bbc6cdac855a6e62d13a37b7a3511/nativelib/src/main/scala/scala/scalanative/runtime/Intrinsics.scala

And my proposal still stand:

- comment every intrinsic function the same way as memset and memcpy
- and make scalafmt happy!

**Comment by fbertra** [comment]
Pending information on note 2:

The allocAtomicAndInit() workaround was for issue https://github.com/scala-native/scala-native/issues/149

**Comment by densh** [comment]
Note 1: I understand the confusion but lets not introduce new terminology for now. We'll probably revisit this name in the future. 

Note 2: The body of `allocAtomicAndInit` should be inlined into `allocAtomic`, given that it's not used anywhere else. Having it in a separate helper doesn't buy us anything. 

Note 3: I think it's best to have a single comment just on the Intrinsics object with a link to general LLVM intrinsics page. It should be quite easy to find corresponding intrinsic given that we use the same names.

**Comment by densh** [comment]
Please address the previous comment, squash all the commits and it's good to go!

**Comment by fbertra** [comment]
@densh done,

I moved allocAtomicAndInit as a nested method inside allocAtomic.  The allocAtomicAndInit method exist only as a workaround of issue https://github.com/scala-native/scala-native/issues/149

**Comment by densh** [comment]
Please squash it all in one commit, but otherwise LGTM!

**Comment by densh** [comment]
#149 has just been fixed btw.

**Comment by fbertra** [comment]
As discussed on gitter, I close this PR to resend a cleaner PR

**Pull request closed without merging**